encoding/ccf/encode.go

```diff
@@ -0,0 +1,1900 @@
+	if err != nil {
+		return err
+	}
+
+	for _, typ := range types {
+
+		switch x := typ.(type) {
+		case cadence.CompositeType:
+			// Encode struct-type, resource-type, contract-type, event-type, or enum-type.
+			err = e.encodeCompositeType(x, tids)
+			if err != nil {
+				return err
+			}
+
+		case cadence.InterfaceType:
+			// Encode struct-interface-type, resource-interface-type, or contract-interface-type.
+			err = e.encodeInterfaceType(x, tids)
+			if err != nil {
+				return err
+			}
+
+		default:
+			panic(cadenceErrors.NewUnexpectedError("unexpected type %s in type definition", typ.ID()))
+		}
+	}
+
+	return nil
+}
+
+// encodeValue traverses the object graph of the provided value and encodes it as
+// language=CDDL
+// value =
+//
+//	ccf-type-and-value-message
+//	/ simple-value
+//	/ optional-value
+//	/ array-value
+//	/ dict-value
+//	/ composite-value
+//	/ path-value
+//	/ capability-value
+//	/ function-value
+//	/ type-value
+//
+// ccf-type-and-value-message =
+//
+//	    ; cbor-tag-type-and-value
+//	    #6.130([
+//	        type: inline-type,
+//		value: value
+//	    ])
+//
+// simple-value =
+//
+//	void-value
+//	/ bool-value
+//	/ character-value
+//	/ string-value
+//	/ address-value
+//	/ uint-value
+//	/ uint8-value
+//	/ uint16-value
+//	/ uint32-value
+//	/ uint64-value
+//	/ uint128-value
+//	/ uint256-value
+//	/ int-value
+//	/ int8-value
+//	/ int16-value
+//	/ int32-value
+//	/ int64-value
+//	/ int128-value
+//	/ int256-value
+//	/ word8-value
+//	/ word16-value
+//	/ word32-value
+//	/ word64-value
+//	/ fix64-value
+//	/ ufix64-value
+//
+// IMPORTANT:
+// "Valid CCF Encoding Requirements" in CCF Specification states:
+//
+//	"Encoders are not required to check for invalid input items
+//	(e.g. invalid UTF-8 strings, duplicate dictionary keys, etc.)
+//	Applications MUST NOT provide invalid items to encoders."
+//
+// cadence.String and cadence.Character must be valid UTF-8
+// and it is the application's responsibility to provide
+// the CCF encoder with valid UTF-8 strings.
+func (e *Encoder) encodeValue(
+	v cadence.Value,
+	staticType cadence.Type,
+	tids ccfTypeIDByCadenceType,
+) error {
+
+	runtimeType := v.Type()
+
+	// CCF requires value to have non-nil type.
+	if runtimeType == nil {
+		panic(cadenceErrors.NewUnexpectedError("value (%T) has nil type", v))
+	}
+
+	if needToEncodeRuntimeType(staticType, runtimeType) {
+		// Get type that needs to be encoded as inline type.
+		inlineType := getTypeToEncodeAsCCFInlineType(staticType, runtimeType)
+
+		// Encode ccf-type-and-value-message.
+
+		// Encode tag number and array head of length 2.
+		err := e.enc.EncodeRawBytes([]byte{
+			// tag number
+			0xd8, CBORTagTypeAndValue,
+			// array, 2 items follow
+			0x82,
+		})
+		if err != nil {
+			return err
+		}
+
+		// element 0: type as inline-type
+		err = e.encodeInlineType(inlineType, tids)
+		if err != nil {
+			return err
+		}
+
+		// element 1: value
+	}
+
+	switch x := v.(type) {
```

Comment:
Can shadow the variable name here:
```suggestion
	switch v := v.(type) {
```

Reply:
Done in commit 926661cbc672ec8711015c5a097586a81e2f3d2b.

While at it, also applied this suggestion to other places.